core/src/connection_reuse.rs

```diff
@@ -40,205 +40,556 @@
+        }
+    }
+
+    /// Polls for the outbound stream of `addr`. Clears the cached value first if `reset` is true.
+    /// Dials, if no connection is in the internal cache. Returns `Ok(Async::NotReady)` as long as
+    /// the connection isn't establed and ready yet.
+    fn poll_outbound(
+        &self,
+        addr: &Multiaddr,
+    ) -> Poll<(SubstreamRef<Arc<C::Output>>, Arc<AtomicBool>), IoError> {
+        let mut conns = self.connections.lock();
+
+        let state = conns.entry(addr.clone()).or_insert_with(|| {
+            let state = match self.transport.clone().dial(addr.clone()) {
+                Ok(future) => {
+                    trace!("Opening new connection to {:?}", addr);
+                    let future = future.and_then(|(out, addr)| addr.map(move |a| (out, a)));
+                    let future = Box::new(future);
+
+                    let mut notify = HashMap::<usize, task::Task, UsizeHasherBuilder>::default();
+                    // make sure we are woken up once this connects
+                    notify.insert(TASK_ID.with(|&t| t), task::current());
+
+                    PeerState::Pending { future, notify }
+                }
+                Err(_) => {
+                    trace!(
+                        "Failed to open connection to {:?}, multiaddr not supported",
+                        addr
+                    );
+                    let err =
+                        IoError::new(IoErrorKind::ConnectionRefused, "multiaddr not supported");
+                    PeerState::Errored(err)
+                }
+            };
+            self.new_con_notify();
+            // Althought we've just started in pending state, there are transports, which are
+            // immediately ready - namely 'memory' - so we need to poll our future right away
+            // rather than waiting for another poll on it.
+            state
+        });
+
+        let (replace_with, return_with) = match state {
+            PeerState::Errored(err) => {
+                // Error: return it
+                let io_err = IoError::new(err.kind(), err.to_string());
+                return Err(io_err);
+            }
+            PeerState::Active {
+                muxer,
+                closed,
+                ..
+            } => {
+                if closed.load(Ordering::Acquire) {
+                    (PeerState::Errored(IoError::new(IoErrorKind::BrokenPipe, "Connection closed")),
+                     Err(IoError::new(IoErrorKind::BrokenPipe, "Connection closed")))
+                } else {
+                    // perfect, let's reuse, update the stream_id and
+                    // return a new outbound
+                    trace!(
+                        "Using existing connection to {} to open outbound substream",
+                        addr
+                    );
+                    match poll_for_substream(muxing::outbound_from_ref_and_wrap(muxer.clone()), &addr) {
```

Comment:
We're creating a new outbound future, but we immediately throw it away if it's not ready? There's no way this can work for something less trivial than mplex.